File: cpp/src/arrow/table/schema-test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <gtest/gtest.h>
#include <memory>
#include <string>
#include <vector>

#include "arrow/table/schema.h"
#include "arrow/type.h"
#include "arrow/types/string.h"

using std::shared_ptr;
using std::vector;

namespace arrow {

TEST(TestField, Basics) {
  shared_ptr<DataType> ftype = INT32;
  shared_ptr<DataType> ftype_nn = std::make_shared<Int32Type>(false);
  Field f0("f0", ftype);
  Field f0_nn("f0", ftype_nn);

  ASSERT_EQ(f0.name, "f0");
  ASSERT_EQ(f0.type->ToString(), ftype->ToString());

  ASSERT_TRUE(f0.nullable());
  ASSERT_FALSE(f0_nn.nullable());
}

TEST(TestField, Equals) {
  shared_ptr<DataType> ftype = INT32;
  shared_ptr<DataType> ftype_nn = std::make_shared<Int32Type>(false);

  Field f0("f0", ftype);
  Field f0_nn("f0", ftype_nn);
  Field f0_other("f0", ftype);

  ASSERT_EQ(f0, f0_other);
  ASSERT_NE(f0, f0_nn);
}

class TestSchema : public ::testing::Test {
 public:
  void SetUp() {}
};

TEST_F(TestSchema, Basics) {
  auto f0 = std::make_shared<Field>("f0", INT32);
  auto f1 = std::make_shared<Field>("f1", std::make_shared<UInt8Type>(false));
  auto f1_optional = std::make_shared<Field>("f1", std::make_shared<UInt8Type>());

  auto f2 = std::make_shared<Field>("f2", std::make_shared<StringType>());

  vector<shared_ptr<Field> > fields = {f0, f1, f2};
  auto schema = std::make_shared<Schema>(fields);

  ASSERT_EQ(3, schema->num_fields());
  ASSERT_EQ(f0, schema->field(0));
  ASSERT_EQ(f1, schema->field(1));
  ASSERT_EQ(f2, schema->field(2));

  auto schema2 = std::make_shared<Schema>(fields);

  vector<shared_ptr<Field> > fields3 = {f0, f1_optional, f2};
  auto schema3 = std::make_shared<Schema>(fields3);
  ASSERT_TRUE(schema->Equals(schema2));
  ASSERT_FALSE(schema->Equals(schema3));

  ASSERT_TRUE(schema->Equals(*schema2.get()));
  ASSERT_FALSE(schema->Equals(*schema3.get()));
}

TEST_F(TestSchema, ToString) {
  auto f0 = std::make_shared<Field>("f0", std::make_shared<Int32Type>());
  auto f1 = std::make_shared<Field>("f1", std::make_shared<UInt8Type>(false));
  auto f2 = std::make_shared<Field>("f2", std::make_shared<StringType>());
  auto f3 = std::make_shared<Field>("f3",
      std::make_shared<ListType>(std::make_shared<Int16Type>()));

  vector<shared_ptr<Field> > fields = {f0, f1, f2, f3};
  auto schema = std::make_shared<Schema>(fields);

  std::string result = schema->ToString();
  std::string expected = R"(f0 int32
f1 uint8 not null
f2 string
f3 list<int16>
)";

  ASSERT_EQ(expected, result);
}

} // namespace arrow

File: go/arrow/avro/schema_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package avro

import (
	"fmt"
	"testing"

	"github.com/apache/arrow/go/v16/arrow"
	hamba "github.com/hamba/avro/v2"
)

func TestSchemaStringEqual(t *testing.T) {
	tests := []struct {
		avroSchema  string
		arrowSchema []arrow.Field
	}{
		{
			avroSchema: `{
				"fields": [
					{
						"name": "inheritNull",
						"type": {
							"name": "Simple",
							"symbols": [
								"a",
								"b"
							],
							"type": "enum"
						}
					},
					{
						"name": "explicitNamespace",
						"type": {
							"name": "test",
							"namespace": "org.hamba.avro",
							"size": 12,
							"type": "fixed"
						}
					},
					{
						"name": "fullName",
						"type": {
							"type": "record",
							"name": "fullName_data",
							"namespace": "ignored",
							"doc": "A name attribute with a fullname, so the namespace attribute is ignored. The fullname is 'a.full.Name', and the namespace is 'a.full'.",
							"fields": [{
									"name": "inheritNamespace",
									"type": {
										"type": "enum",
										"name": "Understanding",
										"doc": "A simple name (attribute) and no namespace attribute: inherit the namespace of the enclosing type 'a.full.Name'. The fullname is 'a.full.Understanding'.",
										"symbols": ["d", "e"]
									}
								}, {
									"name": "md5",
									"type": {
                                            "name": "md5_data",
                                            "type": "fixed",
									        "size": 16,
									        "namespace": "ignored"
                                    }
								}
							]
						}
					},
					{
						"name": "id",
						"type": "int"
					},
					{
						"name": "bigId",
						"type": "long"
					},
					{
						"name": "temperature",
						"type": [
							"null",
							"float"
						]
					},
					{
						"name": "fraction",
						"type": [
							"null",
							"double"
						]
					},
					{
						"name": "is_emergency",
						"type": "boolean"
					},
					{
						"name": "remote_ip",
						"type": [
							"null",
							"bytes"
						]
					},
					{
						"name": "person",
						"type": {
							"fields": [
								{
									"name": "lastname",
									"type": "string"
								},
								{
									"name": "address",
									"type": {
										"fields": [
											{
												"name": "streetaddress",
												"type": "string"
											},
											{
												"name": "city",
												"type": "string"
											}
										],
										"name": "AddressUSRecord",
										"type": "record"
									}
								},
								{
									"name": "mapfield",
									"type": {
										"default": {
										},
										"type": "map",
										"values": "long"
									}
								},
								{
									"name": "arrayField",
									"type": {
										"default": [
										],
										"items": "string",
										"type": "array"
									}
								}
							],
							"name": "person_data",
							"type": "record"
						}
					},
					{
						"name": "decimalField",
						"type": {
							"logicalType": "decimal",
							"precision": 4,
							"scale": 2,
							"type": "bytes"
						}
					},
					{
						"logicalType": "uuid",
						"name": "uuidField",
						"type": "string"
					},
					{
						"name": "timemillis",
						"type": {
							"type": "int",
							"logicalType": "time-millis"
						}
					},
					{
						"name": "timemicros",
						"type": {
								"type": "long",
								"logicalType": "time-micros"
						}
					},
					{
						"name": "timestampmillis",
						"type": {
							"type": "long",
							"logicalType": "timestamp-millis"
						}
					},
					{
						"name": "timestampmicros",
						"type": {
							"type": "long",
							"logicalType": "timestamp-micros"
						}
					},
					{
						"name": "duration",
						"type": {
							"name": "duration",
							"namespace": "whyowhy",
							"logicalType": "duration",
							"size": 12,
							"type": "fixed"
						}
					},
					{
						"name": "date",
						"type": {
							"logicalType": "date",
							"type": "int"
						}
					}
				],
				"name": "Example",
				"type": "record"
			}`,
			arrowSchema: []arrow.Field{
				{
					Name:     "inheritNull",
					Type:     &arrow.DictionaryType{IndexType: arrow.PrimitiveTypes.Uint8, ValueType: arrow.BinaryTypes.String, Ordered: false},
					Metadata: arrow.MetadataFrom(map[string]string{"0": "a", "1": "b"}),
				},
				{
					Name: "explicitNamespace",
					Type: &arrow.FixedSizeBinaryType{ByteWidth: 12},
				},
				{
					Name: "fullName",
					Type: arrow.StructOf(
						arrow.Field{
							Name: "inheritNamespace",
							Type: &arrow.DictionaryType{IndexType: arrow.PrimitiveTypes.Uint8, ValueType: arrow.BinaryTypes.String, Ordered: false},
						},
						arrow.Field{
							Name: "md5",
							Type: &arrow.FixedSizeBinaryType{ByteWidth: 16},
						},
					),
				},
				{
					Name: "id",
					Type: arrow.PrimitiveTypes.Int32,
				},
				{
					Name: "bigId",
					Type: arrow.PrimitiveTypes.Int64,
				},
				{
					Name:     "temperature",
					Type:     arrow.PrimitiveTypes.Float32,
					Nullable: true,
				},
				{
					Name:     "fraction",
					Type:     arrow.PrimitiveTypes.Float64,
					Nullable: true,
				},
				{
					Name: "is_emergency",
					Type: arrow.FixedWidthTypes.Boolean,
				},
				{
					Name:     "remote_ip",
					Type:     arrow.BinaryTypes.Binary,
					Nullable: true,
				},
				{
					Name: "person",
					Type: arrow.StructOf(
						arrow.Field{
							Name:     "lastname",
							Type:     arrow.BinaryTypes.String,
							Nullable: true,
						},
						arrow.Field{
							Name: "address",
							Type: arrow.StructOf(
								arrow.Field{
									Name: "streetaddress",
									Type: arrow.BinaryTypes.String,
								},
								arrow.Field{
									Name: "city",
									Type: arrow.BinaryTypes.String,
								},
							),
						},
						arrow.Field{
							Name:     "mapfield",
							Type:     arrow.MapOf(arrow.BinaryTypes.String, arrow.PrimitiveTypes.Int64),
							Nullable: true,
						},
						arrow.Field{
							Name: "arrayField",
							Type: arrow.ListOfNonNullable(arrow.BinaryTypes.String),
						},
					),
				},
				{
					Name: "decimalField",
					Type: &arrow.Decimal128Type{Precision: 4, Scale: 2},
				},
				{
					Name: "uuidField",
					Type: arrow.BinaryTypes.String,
				},
				{
					Name: "timemillis",
					Type: arrow.FixedWidthTypes.Time32ms,
				},
				{
					Name: "timemicros",
					Type: arrow.FixedWidthTypes.Time64us,
				},
				{
					Name: "timestampmillis",
					Type: arrow.FixedWidthTypes.Timestamp_ms,
				},
				{
					Name: "timestampmicros",
					Type: arrow.FixedWidthTypes.Timestamp_us,
				},
				{
					Name: "duration",
					Type: arrow.FixedWidthTypes.MonthDayNanoInterval,
				},
				{
					Name: "date",
					Type: arrow.FixedWidthTypes.Date32,
				},
			},
		},
	}

	for _, test := range tests {
		t.Run("", func(t *testing.T) {
			want := arrow.NewSchema(test.arrowSchema, nil)
			schema, err := hamba.ParseBytes([]byte(test.avroSchema))
			if err != nil {
				t.Fatalf("%v", err)
			}
			got, err := ArrowSchemaFromAvro(schema)
			if err != nil {
				t.Fatalf("%v", err)
			}
			if !(fmt.Sprintf("%+v", want.String()) == fmt.Sprintf("%+v", got.String())) {
				t.Fatalf("got=%v,\n want=%v", got.String(), want.String())
			} else {
				t.Logf("schema.String() comparison passed")
			}
		})
	}
}
